urdf/src/model.cpp

```diff
@@ -124,59 +148,67 @@ bool Model::initXml(TiXmlElement * robot_xml)
+  try {
+    plugin_instance = loader_.createUniqueInstance(plugin_name);
+  } catch (const pluginlib::CreateClassException &) {
+    fprintf(stderr, "Failed to load urdf_parser_plugin [%s]\n", plugin_name.c_str());
+  }
+  return std::move(plugin_instance);
+}
+
+bool Model::initString(const std::string & data)
 {
   urdf::ModelInterfaceSharedPtr model;
 
-  // necessary for COLLADA compatibility
-  if (IsColladaData(xml_string)) {
-    fprintf(stderr, "Parsing robot collada xml string is not yet supported.\n");
-    return false;
-    /*
-    ROS_DEBUG("Parsing robot collada xml string");
-
-    static boost::mutex PARSER_PLUGIN_LOCK;
-    static boost::scoped_ptr<pluginlib::ClassLoader<urdf::URDFParser> > PARSER_PLUGIN_LOADER;
-    boost::mutex::scoped_lock _(PARSER_PLUGIN_LOCK);
-
-    try
-    {
-      if (!PARSER_PLUGIN_LOADER)
-        PARSER_PLUGIN_LOADER.reset(new pluginlib::ClassLoader<urdf::URDFParser>("urdf_parser_plugin", "urdf::URDFParser"));
-     const std::vector<std::string> &classes = PARSER_PLUGIN_LOADER->getDeclaredClasses();
-     bool found = false;
-     for (std::size_t i = 0 ; i < classes.size() ; ++i)
-       if (classes[i].find("urdf/ColladaURDFParser") != std::string::npos)
-     {
-       boost::shared_ptr<urdf::URDFParser> instance = PARSER_PLUGIN_LOADER->createInstance(classes[i]);
-       if (instance)
-         model = instance->parse(xml_string);
-       found = true;
-       break;
-     }
-     if (!found)
-       ROS_ERROR_STREAM("No URDF parser plugin found for Collada files. Did you install the corresponding package?");
-   }
-   catch(pluginlib::PluginlibException& ex)
-   {
-     ROS_ERROR_STREAM("Exception while creating planning plugin loader " << ex.what() << ". Will not parse Collada file.");
-   }
+  size_t best_score = std::numeric_limits<size_t>::max();
+  pluginlib::UniquePtr<urdf::URDFParser> best_plugin;
+  std::string best_plugin_name;
+
+  // Figure out what plugins might handle this format
+  for (const std::string & plugin_name : impl_->loader_.getDeclaredClasses()) {
+    pluginlib::UniquePtr<urdf::URDFParser> plugin_instance = impl_->load_plugin(plugin_name);
+    if (!plugin_instance) {
+      // Debug mode
+      assert(plugin_instance);
+      // Release mode
+      continue;
+    }
+    size_t score = plugin_instance->might_handle(data);
+    if (score < best_score) {
+      best_score = score;
+      best_plugin = std::move(plugin_instance);
+      best_plugin_name = plugin_name;
+    }
   }
-   */
-  } else {
-    fprintf(stderr, "Parsing robot urdf xml string.\n");
-    model = parseURDF(xml_string);
+
+  if (best_score >= data.size()) {
+    // No plugin was confident ... try urdf anyways
+    best_plugin_name = "urdf_xml_parser/URDFXMLParser";
+    best_plugin = impl_->load_plugin(best_plugin_name);
+  }
+
+  if (!best_plugin) {
+    fprintf(stderr, "No plugin found for given robot description.\n");
+    return false;
   }
```

Comment:
Personally, I'm not sure it makes much sense to always fall back to the URDFXMLParser.  What you've done here covers all of the bases, but I can still imagine a scenario in which the user just doesn't compile URDFXMLParser at all.  It also makes URDFXMLParser "special", which I'm not sure we want.

Could you explain a bit more of your rationale for the fallback?

Reply:
> I can still imagine a scenario in which the user just doesn't compile URDFXMLParser at all

This doesn't seem like a likely scenario to me. This source file is in the same ROS package that [builds the URDF parser extension](https://github.com/ros2/urdf/blob/31014c02b89ba234e13b4ffd8f5d71076720d8aa/urdf/CMakeLists.txt#L52-L54), and if it does the consequences seem small. If someone edits the `CMakeLists.txt` to not build or not install the URDF xml plugin then the consequence (besides the URDF XML parser not being available) is no plugin will be tried by default when no other plugin handles a description.

> Could you explain a bit more of your rationale for the fallback?

It was from feedback from @iantheengineer. When no other plugin wants a description file then no feedback about what's wrong with the file has been printed to the console. Trying to parse it with one of the plugins anyways means the user will get more specific errors about it. Picking URDF XML as the default makes sense in a ROS context where it's the long established format.

> It also makes URDFXMLParser "special", which I'm not sure we want.

If there's going to be a default, it seems reasonable to me to make the URDF XML parser the default parser for the `urdf` package.